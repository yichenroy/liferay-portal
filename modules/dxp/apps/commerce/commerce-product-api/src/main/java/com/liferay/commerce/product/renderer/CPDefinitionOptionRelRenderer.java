/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 * <p>
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * <p>
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.renderer;

import aQute.bnd.annotation.ProviderType;

import com.liferay.portal.kernel.exception.PortalException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * @author Marco Leo
 */
@ProviderType
public interface CPDefinitionOptionRelRenderer {

	public String render(
		long cpDefinitionId, HttpServletRequest httpServletRequest,
		HttpServletResponse httpServletResponse) throws PortalException;

}