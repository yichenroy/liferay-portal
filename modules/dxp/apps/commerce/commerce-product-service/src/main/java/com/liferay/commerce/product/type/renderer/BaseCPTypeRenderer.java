/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.type.renderer;

import com.liferay.commerce.product.type.CPTypeRenderer;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.model.PortletConstants;
import com.liferay.portal.kernel.portlet.PortletBag;
import com.liferay.portal.kernel.portlet.PortletBagPool;
import com.liferay.portal.kernel.util.Portal;
import com.liferay.portal.kernel.util.Validator;
import com.liferay.portal.kernel.util.WebKeys;

import java.io.IOException;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletContext;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
public abstract class BaseCPTypeRenderer implements CPTypeRenderer {

	public void renderJSP(
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		renderJSP(
			getServletContext(httpServletRequest), httpServletRequest,
			httpServletResponse, path);
	}

	public void renderJSP(
			ServletContext servletContext,
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		RequestDispatcher requestDispatcher =
			servletContext.getRequestDispatcher(path);

		try {
			requestDispatcher.include(httpServletRequest, httpServletResponse);
		}
		catch (ServletException se) {
			_log.error("Unable to render JSP " + path, se);

			throw new IOException("Unable to render " + path, se);
		}
	}

	public void setServletContext(ServletContext servletContext) {
		_servletContext = servletContext;
	}

	protected ServletContext getServletContext(
		HttpServletRequest httpServletRequest) {

		if (_servletContext != null) {
			return _servletContext;
		}

		String portletId = portal.getPortletId(httpServletRequest);

		if (Validator.isNotNull(portletId)) {
			String rootPortletId = PortletConstants.getRootPortletId(portletId);

			PortletBag portletBag = PortletBagPool.get(rootPortletId);

			return portletBag.getServletContext();
		}

		return (ServletContext)httpServletRequest.getAttribute(WebKeys.CTX);
	}

	@Reference
	protected Portal portal;

	private static final Log _log = LogFactoryUtil.getLog(
		BaseCPTypeRenderer.class);

	private ServletContext _servletContext;

}