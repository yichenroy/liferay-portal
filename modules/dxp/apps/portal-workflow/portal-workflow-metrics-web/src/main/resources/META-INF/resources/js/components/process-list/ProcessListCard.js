import autobind from 'autobind-decorator';
import EmptyContent from '../../shared/components/EmptyContent';
import gql from 'graphql-tag';
import ProcessListEntries from './ProcessListEntries';
import ProcessListPagination from './ProcessListPagination';
import ProcessListPaginationResults from './ProcessListPaginationResults';
import ProcessListSearch from './ProcessListSearch';
import ProcessListTable from './ProcessListTable';
import React from 'react';

const REQUEST_ORIGIN_TYPE_FETCH = 'REQUEST_ORIGIN_TYPE_FETCH';
const REQUEST_ORIGIN_TYPE_SEARCH = 'REQUEST_ORIGIN_TYPE_SEARCH';

export default class ProcessListCard extends React.Component {
	constructor() {
		super();

		this.state = {
			processes: [],
			requestOriginType: null,
			selectedEntry: 20,
			start: 0,
			total: 0
		};
	}

	componentWillMount() {
		const {selectedEntry, start} = this.state;
		const {client} = this.props;

		const unsubscribe = client.onResetStore(() => {
			this.requestData({size: selectedEntry, start}).then(
				({processes, total}) =>
					this.setState({
						processes,
						total
					})
			);

			unsubscribe();
		});

		client.resetStore();
	}

	requestData({keyword, size, start}) {
		const {client, companyId} = this.props;

		this.state.requestOriginType =
			typeof keyword === 'string'
				? REQUEST_ORIGIN_TYPE_SEARCH
				: REQUEST_ORIGIN_TYPE_FETCH;

		return client
			.query({
				query: gql`
				query workflowProcessesQuery {
					processes(companyId:${companyId}, keyword:"${keyword}", start:${start}, size:${size}) {
						total
						workflowProcesses {
	  						title
	  						instancesCount
						}
					}
				}
			`
			})
			.then(({data: {processes}}) => ({
				processes: processes.workflowProcesses,
				total: processes.total
			}))
			.catch(error => console.error(error));
	}

	@autobind
	onSearch(keyword) {
		const {selectedEntry} = this.state;
		const start = 0;

		this.requestData({keyword, size: selectedEntry, start}).then(
			({processes, total}) => this.setState({processes, start, total})
		);
	}

	@autobind
	setPage({size, start}) {
		return this.requestData({size, start}).then(({processes, total}) =>
			this.setState({processes, start, total})
		);
	}

	@autobind
	setEntry(entry) {
		const start = 0;

		this.requestData({size: entry, start}).then(({processes, total}) =>
			this.setState({processes, selectedEntry: entry, start, total})
		);
	}

	render() {
		const {
			processes,
			requestOriginType,
			selectedEntry,
			start,
			total
		} = this.state;
		const entries = [5, 10, 20, 30, 50, 75];

		const paginationBar = () => {
			if (total > entries[0]) {
				return (
					<div className="pagination-bar">
						<ProcessListEntries
							entries={entries}
							onSelectEntry={this.setEntry}
							selectedEntry={selectedEntry}
						/>

						<ProcessListPaginationResults
							count={processes.length}
							start={start}
							total={total}
						/>

						<ProcessListPagination
							entry={selectedEntry}
							pageClickHandler={this.setPage}
							start={start}
							totalCount={total}
						/>
					</div>
				);
			}
			return <div />;
		};

		return (
			<div>
				<ProcessListSearch disabled onSearch={this.onSearch} />
				<div className="container-fluid-1280">
					{requestOriginType === REQUEST_ORIGIN_TYPE_SEARCH && total === 0 ? (
						<EmptyContent
							message={Liferay.Language.get('no-results-were-found')}
							type="not-found"
						/>
					) : requestOriginType === REQUEST_ORIGIN_TYPE_FETCH && total === 0 ? (
						<EmptyContent
							message={Liferay.Language.get(
								'once-there-are-active-processes-metrics-will-appear-here'
							)}
							title={Liferay.Language.get('no-current-metrics')}
						/>
					) : total > 0 ? (
						<div>
							<ProcessListTable processes={processes} />
							{paginationBar()}
						</div>
					) : (
						<span aria-hidden="true" className="loading-animation" />
					)}
				</div>
			</div>
		);
	}
}