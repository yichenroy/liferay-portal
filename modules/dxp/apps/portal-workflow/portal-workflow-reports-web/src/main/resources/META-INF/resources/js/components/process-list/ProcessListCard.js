import React from 'react';
import ProcessListManagementBar from './ProcessListManagementBar';

export default class ProcessListCard extends React.Component {
	render() {
		return (
			<div className="row">
				<div className="col-12">
					<ProcessListManagementBar />

					<div className="card">
						<div className="card-body">
							<div className="table-responsive">
								<table className="show-quick-actions-on-hover table table-autofit table-hover table-list table-nowrap">
									<thead>
										<tr>
											<th
												className="table-cell-expand table-head-title"
												style={{width: '70%'}}
											>
												{'Process Name'}
											</th>
											<th
												className="table-cell-expand table-head-title"
												style={{width: '15%'}}
											>
												{'On Time'}
											</th>
											<th
												className="text-body table-cell-expand table-head-title"
												style={{width: '15%'}}
											>
												{'Overdue'}
											</th>
										</tr>
									</thead>

									<tbody>
										<tr>
											<td>{'Sales Quoting'}</td>
											<td>{'82 items'}</td>
											<td>{'33 items'}</td>
										</tr>
										<tr>
											<td>{'Refund Request'}</td>
											<td>{'12 items'}</td>
											<td>{'25 items'}</td>
										</tr>
										<tr>
											<td>{'Travel Reimbursement'}</td>
											<td>{'5 items'}</td>
											<td>{'23 items'}</td>
										</tr>
									</tbody>
								</table>
							</div>
						</div>
					</div>
				</div>
			</div>
		);
	}
}