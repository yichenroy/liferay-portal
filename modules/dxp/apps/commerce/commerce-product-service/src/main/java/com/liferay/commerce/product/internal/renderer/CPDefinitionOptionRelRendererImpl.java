/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.internal.renderer;

import com.liferay.commerce.product.model.CPDefinitionOptionRel;
import com.liferay.commerce.product.model.CPDefinitionOptionValueRel;
import com.liferay.commerce.product.renderer.CPDefinitionOptionRelRenderer;
import com.liferay.commerce.product.service.CPDefinitionOptionRelService;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderer;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderingContext;
import com.liferay.dynamic.data.mapping.model.DDMForm;
import com.liferay.dynamic.data.mapping.model.DDMFormField;
import com.liferay.dynamic.data.mapping.model.DDMFormFieldOptions;
import com.liferay.dynamic.data.mapping.model.LocalizedValue;
import com.liferay.portal.kernel.exception.PortalException;
import com.liferay.portal.kernel.util.Portal;

import java.util.List;
import java.util.Locale;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
@Component(immediate = true)
public class CPDefinitionOptionRelRendererImpl
	implements CPDefinitionOptionRelRenderer {

	@Override
	public String render(
			long cpDefinitionId, HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse)
		throws PortalException {

		DDMForm ddmForm = new DDMForm();

		Locale locale = _portal.getLocale(httpServletRequest);

		List<CPDefinitionOptionRel> cpDefinitionOptionRels =
			_cpDefinitionOptionRelService.
				getSkuContributorCPDefinitionOptionRels(cpDefinitionId);

		for (CPDefinitionOptionRel cpDefinitionOptionRel :
				cpDefinitionOptionRels) {

			List<CPDefinitionOptionValueRel> cpDefinitionOptionValueRels =
				cpDefinitionOptionRel.getCPDefinitionOptionValueRels();

			DDMFormField ddmFormField = new DDMFormField(
				String.valueOf(cpDefinitionOptionRel.
					getCPDefinitionOptionRelId()),
				cpDefinitionOptionRel.getDDMFormFieldTypeName());

			if (!cpDefinitionOptionValueRels.isEmpty()) {
				DDMFormFieldOptions ddmFormFieldOptions =
					new DDMFormFieldOptions();

				for (CPDefinitionOptionValueRel cpDefinitionOptionValueRel :
						cpDefinitionOptionValueRels) {

					ddmFormFieldOptions.addOptionLabel(
						String.valueOf(cpDefinitionOptionValueRel.
							getCPDefinitionOptionValueRelId()),
						locale, cpDefinitionOptionValueRel.getTitle(locale));
				}

				ddmFormField.setDDMFormFieldOptions(ddmFormFieldOptions);
			}

			LocalizedValue localizedValue = new LocalizedValue(locale);

			localizedValue.addString(
				locale, cpDefinitionOptionRel.getName(locale));

			ddmFormField.setLabel(localizedValue);

			ddmFormField.setRequired(true);

			ddmForm.addDDMFormField(ddmFormField);
		}

		DDMFormRenderingContext ddmFormRenderingContext =
			new DDMFormRenderingContext();

		ddmFormRenderingContext.setContainerId("options");
		ddmFormRenderingContext.setHttpServletRequest(httpServletRequest);
		ddmFormRenderingContext.setHttpServletResponse(httpServletResponse);
		ddmFormRenderingContext.setLocale(locale);

		return _ddmFormRenderer.render(ddmForm, ddmFormRenderingContext);
	}

	@Reference
	private CPDefinitionOptionRelService _cpDefinitionOptionRelService;

	@Reference
	private DDMFormRenderer _ddmFormRenderer;

	@Reference
	private Portal _portal;

}