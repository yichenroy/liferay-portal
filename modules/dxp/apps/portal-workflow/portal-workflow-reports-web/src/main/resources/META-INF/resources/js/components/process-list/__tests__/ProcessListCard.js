import mockGraphqlClient from '../../../test/mock-graphql-client';
import ProcessListCard from '../ProcessListCard';
import React from 'react';
import renderer from 'react-test-renderer';

test('Should render component', () => {
	mockGraphqlClient.data = {processes: {total: 0, workflowProcesses: []}};

	const component = renderer.create(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});

test('Should render component with 10 records', () => {
	mockGraphqlClient.data = {
		processes: {
			total: 10,
			workflowProcesses: [
				{
					instancesCount: 0,
					title: 'Single Approver 1'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 2'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 3'
				},
				{
					instancesCount: 1,
					title: 'Single Approver 4'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 5'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 6'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 7'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 8'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 9'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 10'
				}
			]
		}
	};

	const component = renderer.create(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});

test('Should render component with 4 records', () => {
	mockGraphqlClient.data = {
		processes: {
			total: 4,
			workflowProcesses: [
				{
					instancesCount: 0,
					title: 'Single Approver 1'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 2'
				},
				{
					instancesCount: 0,
					title: 'Single Approver 3'
				},
				{
					instancesCount: 1,
					title: 'Single Approver 4'
				}
			]
		}
	};

	const component = renderer.create(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});

test('Should change entry', () => {
	mockGraphqlClient.data = {processes: {total: 0, workflowProcesses: []}};

	const component = shallow(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const instance = component.instance();

	instance.setEntry(20);
	expect(component.state('selectedEntry')).toBe(20);
});

test('Should change page', () => {
	mockGraphqlClient.data = {processes: {total: 0, workflowProcesses: []}};

	const component = shallow(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const instance = component.instance();

	instance
		.setPage({size: 10, start: 2})
		.then(() => expect(component.state('start')).toBe(2));
});

test('Should search', () => {
	mockGraphqlClient.data = {processes: {total: 0, workflowProcesses: []}};

	const component = shallow(
		<ProcessListCard client={mockGraphqlClient} companyId={1} />
	);

	const instance = component.instance();

	instance.onSearch('test');
	expect(component.state('total')).toBe(0);
});