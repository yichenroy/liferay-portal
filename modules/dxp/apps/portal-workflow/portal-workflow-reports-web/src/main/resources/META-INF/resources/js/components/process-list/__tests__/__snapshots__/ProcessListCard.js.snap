// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<div
  className="row"
>
  <div
    className="col-12"
  >
    <div>
      <nav
        className="management-bar management-bar-light navbar navbar-expand-md"
      >
        <div
          className="container"
        >
          <ul
            className="navbar-nav"
          >
            <li
              className="dropdown nav-item"
            >
              <a
                aria-expanded="false"
                aria-haspopup="true"
                className="dropdown-toggle nav-link navbar-breakpoint-down-d-none"
                data-toggle="dropdown"
                href="#1"
                role="button"
              >
                <span
                  className="navbar-text-truncate"
                >
                  Filter and Order2
                </span>
                <svg
                  className="lexicon-icon lexicon-icon-caret-bottom"
                  focusable="false"
                  role="presentation"
                >
                  <use
                    href="/images/icons/icons.svg#caret-bottom"
                  />
                </svg>
              </a>
              <a
                aria-expanded="false"
                aria-haspopup="true"
                className="nav-link nav-link-monospaced dropdown-toggle navbar-breakpoint-d-none"
                data-toggle="dropdown"
                href="#1"
                role="button"
              >
                <svg
                  className="lexicon-icon lexicon-icon-filter"
                  focusable="false"
                  role="presentation"
                >
                  <use
                    href="/images/icons/icons.svg#filter"
                  />
                </svg>
              </a>
              <div
                className="dropdown-menu"
                role="menu"
              >
                <ul
                  className="list-unstyled"
                >
                  <li>
                    <a
                      className="dropdown-item"
                      href="#1"
                      role="button"
                    >
                      Filter Action 1
                    </a>
                  </li>
                  <li>
                    <a
                      className="dropdown-item"
                      href="#1"
                      role="button"
                    >
                      Filter Action 2
                    </a>
                  </li>
                  <li>
                    <a
                      className="dropdown-item"
                      href="#1"
                      role="button"
                    >
                      Filter Action 3
                    </a>
                  </li>
                </ul>
              </div>
            </li>
          </ul>
          <div
            className="navbar-form navbar-form-autofit navbar-overlay navbar-overlay-sm-down"
          >
            <div
              className="container"
            >
              <form
                role="search"
              >
                <div
                  className="input-group"
                >
                  <div
                    className="input-group-item"
                  >
                    <input
                      className="form-control input-group-inset input-group-inset-after"
                      placeholder="Search for..."
                      type="text"
                    />
                    <span
                      className="input-group-inset-item input-group-inset-item-after"
                    >
                      <button
                        className="btn btn-unstyled navbar-breakpoint-d-none"
                        type="button"
                      >
                        <svg
                          className="lexicon-icon lexicon-icon-times"
                          focusable="false"
                          role="presentation"
                        >
                          <use
                            href="/images/icons/icons.svg#times"
                          />
                        </svg>
                      </button>
                      <button
                        className="btn btn-unstyled navbar-breakpoint-d-block"
                        type="button"
                      >
                        <svg
                          className="lexicon-icon lexicon-icon-search"
                          focusable="false"
                          role="presentation"
                        >
                          <use
                            href="/images/icons/icons.svg#search"
                          />
                        </svg>
                      </button>
                    </span>
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </nav>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="table-responsive"
        >
          <table
            className="show-quick-actions-on-hover table table-autofit table-hover table-list table-nowrap"
          >
            <thead>
              <tr>
                <th
                  className="table-cell-expand table-head-title"
                  style={
                    Object {
                      "width": "70%",
                    }
                  }
                >
                  Process Name
                </th>
                <th
                  className="table-cell-expand table-head-title"
                  style={
                    Object {
                      "width": "15%",
                    }
                  }
                >
                  On Time
                </th>
                <th
                  className="text-body table-cell-expand table-head-title"
                  style={
                    Object {
                      "width": "15%",
                    }
                  }
                >
                  Overdue
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  Sales Quoting
                </td>
                <td>
                  82 items
                </td>
                <td>
                  33 items
                </td>
              </tr>
              <tr>
                <td>
                  Refund Request
                </td>
                <td>
                  12 items
                </td>
                <td>
                  25 items
                </td>
              </tr>
              <tr>
                <td>
                  Travel Reimbursement
                </td>
                <td>
                  5 items
                </td>
                <td>
                  23 items
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
